src/search/Search.tsx

```diff
@@ -1,53 +1,51 @@
 import { useState, useEffect, useRef, useCallback } from "react";
 import { Action } from "../core/Action";
-import { SearchController } from "./SearchController";
 
 export interface SearchState {
     selectedIndex: number,
     query: string
 }
 
-function selectNext(state: SearchState, results: Action[]): SearchState {
-    var index = state.selectedIndex + 1;
-    if (index >= results.length) {
-        index = results.length-1;
-    }
-    return {...state, selectedIndex: index};
-}
-
-function selectPrev(state: SearchState): SearchState {
-    var index = state.selectedIndex - 1;
-    if (index < 0) {
-        index = 0;
-    }
-    return {...state, selectedIndex: index};
-}
-
 interface SearchProps {
-    onClose?: () => void,
-    controller: SearchController
+    initialQuery: string,
+    results: Action[],
+    prompt: string,
+    onCloseRequest: () => void,
+    onSearch: (query: string) => void,
 }
 
-export function Search({ controller, onClose }: SearchProps) {
-    const _onClose = onClose || (() => { controller.hideSearch() });
-    const _onSearch = (query: string) => { controller.search(query) };
-    const _initialQuery = controller.initialQuery;
-    const results = controller.searchResults;
-    const visible = controller.searchVisible;
+export function Search({ 
+    initialQuery = "" ,
+    results = [],
+    prompt = "",
+    onCloseRequest = ()=>{},
+    onSearch = (_: string)=>{},
+}: SearchProps) {
 
-    const [state, updateState] = useState<SearchState>({ query: _initialQuery, selectedIndex: 0 });
+    const [state, updateState] = useState<SearchState>({ query: initialQuery, selectedIndex: 0 });
     const inputRef = useRef<HTMLInputElement>(null);
 
     useEffect(() => {
-        // focus the input when the dialog is shown
-        if (visible && inputRef.current) {
+        if (inputRef.current) {
             inputRef.current.focus();
         }
-        // reset the query when the dialog is hidden
-        if (!visible) {
-            updateState({ query: "", selectedIndex: 0 });
+    }, [inputRef]);
+
+    function selectNext(state: SearchState, results: Action[]): SearchState {
+        var index = state.selectedIndex + 1;
+        if (index >= results.length) {
+            index = results.length - 1;
         }
-    }, [visible]);
+        return { ...state, selectedIndex: index };
+    }
+
+    function selectPrev(state: SearchState): SearchState {
+        var index = state.selectedIndex - 1;
+        if (index < 0) {
+            index = 0;
+        }
+        return { ...state, selectedIndex: index };
+    }
 
     const handleKeyPress = useCallback((event: KeyboardEvent) => {
```

Comment:
wydaje mi się, że useCallback jest zbędny

Reply:
A nie warto tutaj zrobić jakiegoś useMemo żeby handleKeyPress nie była generowana przy każdym renderowaniu?

Reply:
Wydaje mi się, że w tym przypadku nie ma takiej potrzeby :) Hooki optymalizacyjne takie jak useMemo i useCallback same są pamięciożerne i z mojego doświadczenia raczej sięga się po nie dopiero kiedy pojawia się problem wydajnościowy (albo kiedy mamy świadomość, że taki może nastąpić), a nie do rozwiązywania hipotetycznych problemów - polecam artykuł Kenta (frontowy guru) na ten temat https://kentcdodds.com/blog/usememo-and-usecallback :)

Reply:
Super! Poczytam, dzięki :)